<!DOCTYPE html>
<!--
template.html
Jinja template for Tapestry report

Part of Tapestry
https://github.com/johnomics/tapestry

MIT License

Copyright (c) 2019 John Davey

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
-->


<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="Tapestry: quality assessment for small genome assemblies">
  <meta name="author" content="Tapestry">

  <title>Tapestry Report</title>
  
  <style type="text/css">
      {{ include_file('static/litera.epic.css') }}
      {{ include_file('static/bootstrap-table.min.css') }}
  </style>
  <style>
      html {
          font-size: 14px;
      }
      .section {
          display: flex;
      }
      body {
          padding-top: 50px;
      }
      .footer {
        width: 100%;
        height: 60px; /* Set the fixed height of the footer here */
        line-height: 60px; /* Vertically center the text there */
        background-color: #f5f5f5;
      }
      .alert {
          display: none;
      }
      :target::before {
        content: "";
        display: block;
        height: 50px; /* fixed header height*/
        margin: -50px 0 0; /* negative fixed header height */
      }
      .xaxis { font-size: 14px; }
      .yaxis { font-size: 14px; }
      .xaxislabel { font-size: 16px; font-weight: bold; }
      .yaxislabel { font-size: 16px; font-weight: bold; }
  }
  </style>
</head>

<body>
  <div class="mainpage" style="margin-left: 10px">
    <noscript>
      <div class="alert alert-danger">
        <h4>JavaScript Disabled</h4>
        <p>
          Tapestry uses JavaScript, but it looks like
          you have JavaScript disabled in your web browser.
          Please enable JavaScript to view the full report.
        </p>
      </div>
    </noscript>

    <nav class="navbar navbar-expand-sm fixed-top navbar-dark bg-primary">
      <a class="navbar-brand" href="https://github.com/johnomics/tapestry">Tapestry Report</a>
      <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#collapsingNavbar">
          <span class="navbar-toggler-icon"></span>
      </button>
      <div class="navbar-collapse collapse" id="collapsingNavbar">
        <ul class="navbar-nav">
          <li class="nav-item active">
            <a class="nav-link" href="#assembly">Assembly</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#contigs">Contigs</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#reads">Reads</a>
          </li>
          <li class="nav-item">
            <a class="nav-link" href="#information">Information</a>
          </li>
        </ul>
      </div>
    </nav>
    
    <br>
    <p>
        This is a Tapestry report on the genome assembly <strong>{{assemblyfile}}</strong>. See <a href="#information">Information</a> for full details.
    </p>
    <div>
      If you have a contig order in a Tapestry CSV file, load it here:
      <input type="file" id="assembly_csv" name="assembly_csv" accept=".csv, text/csv" aria-describedby="filehelp">
      <div class="alert alert-warning badcsv" role="alert">
          Contigs in CSV file do not match contigs in report. Please choose a CSV file that matches this assembly.
          <button type="button" class="close" aria-hidden="true">&times;</button>
      </div>
      <div class="alert alert-success goodcsv" role="alert">
          Contigs loaded successfully!
          <button type="button" class="close" aria-hidden="true">&times;</button>
      </div>
    </div>
    
    <hr/>
    <h2><a name="assembly">Assembly</a></h2>
    <p>Selected and rejected contigs are those checked and unchecked in the <a href="#contigs">contig table</a> <strong>Keep</strong> column below.</p>
    <div class="section">
      <table id="assembly_table" style="width: auto" data-sort-order="asc">
        <thead>
          <tr>
            <th data-field="assembly" data-sortable="true">Contig set</th>
            <th data-field="contigs" data-align='right' data-sortable="true">Contig number</th>
            <th data-field="length" data-align='right' data-sortable="true">Length (bp)</th>
            <th data-field="n50_number" data-align='right' data-sortable="true">N50 number</th>
            <th data-field="n50_length" data-align='right' data-sortable="true">N50 length</th>
          </tr>
        </thead>
      </table>
    </div>
    <hr/>
    
    <div style="margin-top:20px">
      <h2><a name="contigs">Contigs</a></h2>
      <p><span style="color:crimson">Red vertical lines</span> at the ends of contigs show telomeres. Opacity of red lines reflects the number of telomere sequences found at each contig end. <span style="color:green">Green blocks</span> show ploidy levels; see <a href="#reads">read alignment plot</a> for key.</p>
      <p>Click on a table radio button to show <a href="#reads">read alignments</a> for the selected contig.</p>
      <p>Click on plot contig names to show contig alignments. Hover over <span style="color:blue; opacity:0.3">blue alignment blocks</span> to show connecting alignments. The selected contig is shown with a dashed black line. <span style="color:purple">Purple lines</span> show contigs with alignments to the selected contig. The opacity of each purple line shows the percentage of the selected contig found in the aligned contig. Dark lines suggest the selected contig is mostly contained in the aligned contig.
      <div class="section">
          <div id="contigs_plot"></div>
          <table id="contig_table">
            <thead>
              <tr>
                <th data-radio="true"></th>
                <th data-field="name" data-sortable="true" data-align='left'>Contig</th>
                <th data-field="length" data-sortable="true" data-align='right' data-formatter="local_formatter">Length</th>
                <th data-field="gc" data-sortable="true" data-formatter="gc_formatter" data-align='right'>GC %</th>
                <th data-field="median_read_depth" data-sortable="true" data-formatter="depth_formatter" data-align='right'>Read Depth</th>
                <th data-field="group" data-sortable="true" data-editable="true" data-align='right'>Group</th>
                <th data-field="included" data-sortable="true" data-formatter="include_checkbox" data-align='center'>Keep</th>
                <th data-field="notes" data-sortable="true" data-editable="true">Notes</th>
              </tr>
            </thead>
          </table>
      </div>
    </div>
    <hr/>
    
    <h2><a name="reads">Read Alignments</a></h2>
    <div class="section">
      <div id="read_alignment_plot"></div>
      <div id="read_alignment_text" style="margin-left: 20px; margin-top: 65px; width: 550px;">
        <p>The contig selected by radio button in the <a href="#contigs">contig table</a> is shown.</p>
        <p>Read alignments are shown in <span style="color:blue">blue</span>.</p>
        <p>Clipped parts of reads with no alignments are shown in <span style="color:wheat">beige</span>.</p>
        <p>Where parts of reads align elsewhere, lengths of unaligned read sequence connecting to neighbouring alignments are shown for alignments to the <br><span style="color:slategray">same contig</span> or <span style="color:mediumorchid">different contigs</span>.</p>
        <p>Hover over alignments to show names of connecting contigs (or '-' if none).</p>
        <p>Alignment opacity shows mapping quality; darker alignments are higher quality.</p>
      </div>
    </div>
    <hr/>
    
    <h2><a name="information">Information</a></h2>
    <div class="section" style="margin-bottom: 20px">
      <table id="options_table" data-show-header="false" style="margin-top: 20px">
        <thead>
          <tr>
            <th data-width="200px" data-field="option"  data-formatter="info_formatter">Option</th>
            <th data-field="value" data-formatter="local_formatter">Value/th>
          </tr>
        </thead>
      </table>
    </div>
  </div>


  <footer class="footer">
      Â© 2019-2020 <a href="https://github.com/johnomics/tapestry/">John Davey</a>.
      Built with <a href="https://d3js.org">D3</a>,
      <a href="https://getbootstrap.com">Bootstrap</a>,
      <a href="https://bootstrap-table.com">Bootstrap Table</a>,
      <a href="https://vitalets.github.io/x-editable/">X-editable</a>,
      <a href="https://github.com/hhurz/tableExport.jquery.plugin">TableExport</a>,
      <a href="https://github.com/eligrey/FileSaver.js">FileSaver.js</a>,
      <a href="https://jquery.com">JQuery</a>,
      <a href="https://www.papaparse.com">PapaParse</a>,
      <a href="https://popper.js.org">Popper.js</a>, and the 
      <a href="https://bootswatch.com/litera/">Litera</a> theme via 
      <a href="https://epicbootstrap.com">Epic Bootstrap</a>.
  </footer>

<script type="text/javascript">
{{ include_file('static/jquery-3.3.1.min.js') }}
{{ include_file('static/popper.min.js') }}
{{ include_file('static/bootstrap.min.js') }}
{{ include_file('static/bootstrap-editable.min.js') }}
{{ include_file('static/FileSaver.min.js') }}
{{ include_file('static/tableExport.min.js') }}
{{ include_file('static/fontawesome.all.min.js') }}
{{ include_file('static/bootstrap-table.min.js') }}
{{ include_file('static/bootstrap-table-export.min.js') }}
{{ include_file('static/bootstrap-table-editable.min.js') }}
{{ include_file('static/papaparse.min.js') }}

// Use hacked version for Bootstrap 4 from https://github.com/wenzhixin/bootstrap-table/issues/4258
{{ include_file('static/bootstrap-table-multiple-sort.js') }}

{{ include_file('static/d3.min.js') }}

var windowsize = {{ windowsize }};
var outdir = "{{ outdir }}";
var options = {{ options }};
var median_depth = {{ median_depth }};
var contigs = {{ contigs }};
var read_alignments = {{ read_alignments }};
var contig_alignments = {{ contig_alignments }};
var contig_coverage = {{ contig_coverage }};
var ploidys = {{ ploidys }};

var current_aligned_contig = '';

ploidy_depths = [ {name:'none',       depth: 0                 },
                  {name:'haploid',    depth: median_depth * 0.5}, 
                  {name:'diploid',    depth: median_depth      }, 
                  {name:'triploid',   depth: median_depth * 1.5}, 
                  {name:'tetraploid', depth: median_depth * 2  },
                  {name:'repeat',     depth: median_depth * 2.5} ];

var contig_names = {};
var contig_id_name = {};
var contig_name_id = {};
var contig_lengths = {};
var max_contig_length = 0;
contigs.forEach(function(contig) {
    contig.included = 1;
    contig.notes = '';
    contig_names[contig.name] = contig.name;
    contig_id_name[contig.id] = contig.name;
    contig_name_id[contig.name] = contig.id;
    contig_lengths[contig.id] = contig.length;
    if (contig.length > max_contig_length) {
        max_contig_length = contig.length;
    }
})

var unit = 'bp';
if (max_contig_length > 2000000) {
    unit = "Mb";
}

var contig_links = {};
for (var contig_name in contig_alignments) {
    contig_links[contig_name] = [];
    contig_matches = [...new Set(contig_alignments[contig_name].map(d => contig_id_name[d[2]]))]
    contig_matches.forEach(function(contig_match) {
        if (contig_name != contig_match) {
            contig_links[contig_name].push({
                source: {"x":-1, "y": contig_name},
                target: {"x":0, "y": contig_match},
                self: false
            });
        }
    })
    contig_links[contig_name].push({
        source: {"x":-1, "y": contig_name},
        target: {"x":0, "y": contig_name},
        self: true
    });
}

var contig_coverage_prop = {};
for (var contig_id_1 in contig_coverage) {
    contig_coverage_prop[contig_id_1] = {};
    for (var contig_id_2 in contig_coverage[contig_id_1]) {
        contig_coverage_prop[contig_id_1][contig_id_2] = contig_coverage[contig_id_1][contig_id_2]/contig_lengths[contig_id_1];
    }
}

function local_formatter(value) {
    return value.toLocaleString();
}

function gc_formatter(value) {
    return '<div style="color: ' + assembly.dim.gcscale(value/100) + '">' + value + '</div>';
}

function depth_formatter(value) {
    return '<div style="color: ' + assembly.dim.depthscale(0.35 + (0.65*value)/median_depth) + '">' + value + '</div>';
}

function info_formatter(value) {
    return '<strong>' + value + '</strong>';
}

function mq_opacity(mq) { return 0.1 + 0.9*mq/60 }

function ploidy_colour(assembly, ploidy) { return assembly.dim.ploidyscale(0.01 + ploidy/5 * 0.99); }

function isEmpty(obj) {
  return Object.keys(obj).length === 0;
}

function get_assembly_length(contigs) {
    return contigs.reduce((a,b) => ({length: a.length + b.length}), {length: 0}).length;
}

function check_max(max_val, list, i) {
    var list_max = d3.max(list, function(d) {return d[i]});
    if (max_val < list_max) {
        max_val = list_max;
    }
    return max_val;
}

function check_min(min_val, list, i) {
    var list_min = d3.min(list, function(d) {return d[i]});
    if (min_val > list_min) {
        min_val = list_min;
    }
    return min_val;
}

function get_n50_stats(contiglist, assembly_length) {
    var n50_length = 0
    var n50_number = 0
    var n50_cumul_length = 0
    
    contiglist.slice(0).sort((a,b) => b.length - a.length).forEach(function(c) {
        if (n50_cumul_length >= assembly_length / 2) { return; }
        n50_number += 1;
        n50_length = c.length;
        n50_cumul_length += c.length;
    })
    
    return [n50_number, n50_length];
}

function calculate_assembly_stats(contigs, label) {
    var assembly_length = get_assembly_length(contigs);
    var [n50_number, n50_length] = get_n50_stats(contigs, assembly_length);

    return {assembly: label,
            contigs: contigs.length,
            length: assembly_length.toLocaleString(),
            n50_number: n50_number,
            n50_length: n50_length.toLocaleString()};
}

function calculate_stats(contigs) {
    return [
        calculate_assembly_stats(contigs, "All contigs"),
        calculate_assembly_stats(contigs.slice(0).filter(c => c.included), "Selected contigs"),
        calculate_assembly_stats(contigs.slice(0).filter(c => !c.included), "Rejected contigs")
    ]
}

function clip_colour(type) {
    if (type === 'C') { // Connection
        return 'mediumorchid'
    } else if (type === 'L') { // Loose
        return 'wheat'
    } else if (type === 'S') { // Self (within-contig)
        return 'slategray'
    } else {
        return 'black'
    }
}

function contig_text_hover(x, y, contig_id, anchor, dim, reads) {
    reads.svg.append("text")
        .attr("class", "alignment text")
        .attr("x", dim.xscale(x))
        .attr("y", reads.dim.yscale(y))
        .attr("text-anchor", anchor)
        .attr("alignment-baseline", "middle")
        .attr("fill", "black")
        .attr("font-size", "10px")
        .text(contig_id == -1 ? '-' : contig_id_name[contig_id]);
}

function contig_text_unhover() {
    reads.svg.selectAll("text.alignment.text").remove()
}

function build_assembly_plot(contigs, dim, windowsize, xaxis, unit) {

    var assembly = {};
    assembly.dim = {height: contigs.length * 29};
    
    assembly.dim.yscale = d3.scaleBand()
                          .domain(contigs.map(d => d.name))
                          .rangeRound([0, assembly.dim.height])
                          .paddingInner(0.8)
                          .paddingOuter(0.4);

    assembly.dim.ploidyscale = d3.scaleSequential(d3.interpolateGreens);
    assembly.dim.gcscale = d3.scaleSequential(d3.interpolateRainbow);
    assembly.dim.depthscale = d3.scaleSequential(d3.interpolateGreys);

    assembly.svg = d3.select("#contigs_plot")
                .append("svg")
                .attr("width", dim.width + dim.margin.left + dim.margin.right/2)
                .attr("height", assembly.dim.height + dim.margin.top + dim.margin.bottom)
                .append("g")
                .attr("transform", "translate(" + dim.margin.left + "," + (dim.margin.top + assembly.dim.height) + ")")
                .call(xaxis)
                .attr("class", "xaxis")
                .append("g")
                .attr("transform", "translate(0,-" + assembly.dim.height + ")")
                .call(d3.axisLeft(assembly.dim.yscale))
                .attr("class", "yaxis");

    assembly.svg
        .selectAll(".tick")
        .on("click", function(d) { show_contig_alignments(d, assembly); })
    
    assembly.svg.append("text")
        .attr("class", "xaxislabel")
        .attr("transform", "translate(" + (dim.width/2) + " ," + (assembly.dim.height + 45) + ")")
        .style("text-anchor", "middle")
        .attr("fill", "black")
        .text("Position (" + unit + ")");
    
    assembly.svg.selectAll("g.contig")
        .data(contigs, function(d) { return d.name })
        .enter()
        .append("g")
        .attr("class", "contig")
        .attr("name", function(d) { return d.name })
        .attr("length", function(d) { return d.length })
        .each(function(d) { build_contig(d3.select(this), dim, assembly, windowsize); });

    return assembly;
}

function build_reads_plot(read_alignments, dim, max_read_depth, assembly, ploidy_depths, xaxis) {

    var reads = {};
    
    reads.dim = { height: 300 };

    reads.dim.yscale = d3.scaleLinear()
                         .domain([0, d3.min([max_read_depth, median_depth * 4])])
                         .range([reads.dim.height, 0]);

    reads.svg = d3.select("#read_alignment_plot")
                  .append("svg")
                  .attr("width", dim.width + dim.margin.left + dim.margin.right)

    reads.svg = reads.svg
        .attr("height", reads.dim.height + 20 + dim.margin.bottom)
        .append("g")
        .attr("transform", "translate(" + dim.margin.left + "," + (20 + reads.dim.height) + ")")
        .call(xaxis)
        .attr("class", "xaxis")
        .append("g")
        .attr("transform", "translate(0,-" + reads.dim.height + ")")
        .call(d3.axisLeft(reads.dim.yscale))
        .attr("class", "yaxis")
    
    reads.svg.append("text")
        .attr("class", "yaxislabel")
        .attr("transform", "rotate(-90)")
        .attr("y", 0 - (dim.margin.left * 0.6))
        .attr("x",0 - (reads.dim.height / 2))
        .attr("fill", "black")
        .style("text-anchor", "middle")
        .text("Reads"); 

    reads.svg.selectAll("line.ploidy")
        .data(ploidy_depths.filter(d => d.name != "repeat" && d.name != "none"))
        .enter()
        .append("line")
        .attr("class", "ploidy")
        .attr("x1", dim.xscale.range()[0]+1) // +1 to avoid axis
        .attr("x2", dim.xscale.range()[1])
        .attr("y1", function(d) { return reads.dim.yscale(d.depth) })
        .attr("y2", function(d) { return reads.dim.yscale(d.depth) })
        .style("stroke", function(d, i) { return ploidy_colour(assembly, i+1); }) //+1 because none is filtered out
        .style("stroke-width", 2)
    
    reads.svg.selectAll("text.ploidy")
        .data(ploidy_depths)
        .enter()
        .append("text")
        .attr("x", dim.xscale.range()[1] + 10 )
        .attr("y", function(d) { return reads.dim.yscale(d.depth)+2 })
        .style("fill", function(d, i) { return ploidy_colour(assembly, i); })
        .style("text-anchor", "start")
        .text(function(d) { return d.name })
    
    return reads;
}

function plot_read_alignments(alignments, dim, reads, contig, unit) {

    // contig_read_alignments:
    //  0 read start including left clip
    //  1 contig alignment start
    //  2 contig alignment end
    //  3 read end including right clip
    //  4 mapping quality
    //  5 y position on plot

    reads.svg.selectAll("text.xaxislabel").remove()
    
    reads.svg.append("text")
       .attr("class", "xaxislabel")
       .attr("transform", "translate(" + (dim.width/2) + " ," + (reads.dim.height + 45) + ")")
       .style("text-anchor", "middle")
       .attr("fill", "black")
       .text(contig + " Position (" + unit + ")");

    reads.svg.selectAll("line.contig_boundary").remove()
       
    reads.svg.append("line")
       .attr("class", "contig_boundary")
       .attr("x1", dim.xscale(0) )
       .attr("x2", dim.xscale(0) )
       .attr("y1", reads.dim.yscale.range()[0] )
       .attr("y2", reads.dim.yscale.range()[1] )
       .style("stroke-dasharray", "5,5")
       .style("stroke", "black")

    reads.svg.append("line")
       .attr("class", "contig_boundary")
       .attr("x1", dim.xscale(contig_lengths[contig_name_id[contig]]) )
       .attr("x2", dim.xscale(contig_lengths[contig_name_id[contig]]) )
       .attr("y1", reads.dim.yscale.range()[0] )
       .attr("y2", reads.dim.yscale.range()[1] )
       .style("stroke-dasharray", "5,5")
       .style("stroke", "black")

    if (alignments === undefined) {
        reads.svg = reads.svg.append("text")
            .attr("x", dim.xscale(0))
            .attr("y", reads.dim.yscale.range()[1])
            .style("text-anchor", "start")
            .attr("fill", "black")
            .text("Read alignments are not available in this report (re-run Tapestry with the -f option to include them).")
        return;
    }

    reads.svg.selectAll("line.read_alignment").remove()

    var ra = reads.svg.selectAll("line.read_alignment")
        .data(alignments)
        .enter();
    
    ra.append("line")
        .attr("class", "read_alignment main")
        .attr("x1", function(d) { return dim.xscale(d[1]); })
        .attr("x2", function(d) { return dim.xscale(d[2]); })
        .attr("y1", function(d) { return reads.dim.yscale(d[5]); })
        .attr("y2", function(d) { return reads.dim.yscale(d[5]); })
        .style("stroke-opacity", function(d) { return mq_opacity(d[4]) })
        .style("stroke", "blue")
        .on("mouseover", function(d) {
            contig_text_hover(d[0], d[5], d[6], "end", dim, reads);
            contig_text_hover(d[3], d[5], d[7], "start", dim, reads);
        })
        .on("mouseout", contig_text_unhover);

    ra.append("line")
        .attr("class", "read_alignment clip left")
        .attr("x1", function(d) { return dim.xscale(d[0]); })
        .attr("x2", function(d) { return dim.xscale(d[1]); })
        .attr("y1", function(d) { return reads.dim.yscale(d[5]+0.1); })
        .attr("y2", function(d) { return reads.dim.yscale(d[5]+0.1); })
        .style("stroke-opacity", function(d) { return mq_opacity(d[4]); })
        .style("stroke", function(d) { return clip_colour(d[8]); })
        .on("mouseover", function(d) {
            contig_text_hover(d[0], d[5], d[6], "end", dim, reads);
        })
        .on("mouseout", contig_text_unhover);

    ra.append("line")
          .attr("class", "read_alignment clip right")
          .attr("contig_id", function(d) { return d[7]})
          .attr("x1", function(d) { return dim.xscale(d[2]); })
          .attr("x2", function(d) { return dim.xscale(d[3]); })
          .attr("y1", function(d) { return reads.dim.yscale(d[5]+0.1); })
          .attr("y2", function(d) { return reads.dim.yscale(d[5]+0.1); })
          .style("stroke-opacity", function(d) { return mq_opacity(d[4]) })
          .style("stroke", function(d) { return clip_colour(d[9])})
        .on("mouseover", function(d) {
            contig_text_hover(d[3], d[5], d[7], "start", dim, reads);
        })
        .on("mouseout", contig_text_unhover);

}

function connect_alignments(d, i) {
    this_y = assembly.dim.yscale(d3.select(this).attr("contig1"))
    other_y = assembly.dim.yscale(d3.select(this).attr("contig2"))

    if (this_y == other_y) {
        this_y += assembly.dim.yscale.bandwidth();
        other_y += assembly.dim.yscale.bandwidth()*2;
        xpoints = [d[0], d[1], d[3], d[4]].sort((a,b) => Number(a) - Number(b));
        x1 = xpoints[1]; // Make sure inner points are used,
        x2 = xpoints[1]; // so new rect doesn't overlap
        x3 = xpoints[2]; // existing rects
        x4 = xpoints[2];
    } else {
        if (this_y < other_y) {
            this_y += assembly.dim.yscale.bandwidth()*2;
            other_y += assembly.dim.yscale.bandwidth();
        } else {
            this_y += assembly.dim.yscale.bandwidth();
            other_y += assembly.dim.yscale.bandwidth()*2;
        }
        x1 = d[0];
        x2 = d[3];
        x3 = d[4];
        x4 = d[1];
    }
    
    var poly = [{"x":dim.xscale(x1), "y":this_y },
                {"x":dim.xscale(x2), "y":other_y },
                {"x":dim.xscale(x3), "y":other_y },
                {"x":dim.xscale(x4), "y":this_y }];
    
    assembly.svg.selectAll("polygon.connector")
        .data([poly])
        .enter()
        .append("polygon")
        .attr("class", "connector")
        .attr("points",function(d) { 
               return d.map(function(d) {
                   return [d.x,d.y].join(",");
               }).join(" ");
           })
        .attr("fill", "purple")
        .attr("fill-opacity",0.1);
}

function disconnect_alignments(d, i) {
    assembly.svg.selectAll(".connector").remove()
}

function show_contig_alignments(contig, assembly) {

    assembly.svg.selectAll("path.contig_link").remove();
    assembly.svg.selectAll("rect.contig_alignment").remove();

    if (contig == current_aligned_contig) {
        current_aligned_contig = '';
        return;
    }

    current_aligned_contig = contig;
    
    assembly.svg.selectAll("path.contig_link")
        .data(contig_links[contig])
        .enter()
        .append("path")
        .attr("class", "contig_link")
        .attr("d", d3.linkHorizontal()
                    .x(function(d) { return d.x == -1 ? 0 : dim.xscale(d.x); })
                    .y(function(d) { return assembly.dim.yscale(d.y) + assembly.dim.yscale.bandwidth()/2; })
        )
        .style("stroke-opacity", function(d) {
            return d.self ? 1 : 0.1 + 0.9*contig_coverage_prop[contig_name_id[contig]][contig_name_id[d.target.y]];
        })
        .style("stroke", function(d) { return d.self ? "black" : "purple"; })
        .style("stroke-dasharray", function(d) { return d.self ? "2,2" : "none" });

    assembly.svg.selectAll("rect.contig_alignment.self")
        .data(contig_alignments[contig])
        .enter()
        .append("rect")
        .attr("class", "contig_alignment self")
        .attr("contig1", contig)
        .attr("contig2", function(d) { return contig_id_name[d[2]]; })
        .attr("x", function(d) { return dim.xscale(d[0]); })
        .attr("width", function(d) { return dim.xscale(d[1])-dim.xscale(d[0]); })
        .attr("y", function(d) { return assembly.dim.yscale(contig) + assembly.dim.yscale.bandwidth(); })
        .attr("height", function(d) { return assembly.dim.yscale.bandwidth(); })
        .style("fill", "blue")
        .style("fill-opacity", 0.1)
        .style("stroke-opacity", 0)
        .on("mouseover", connect_alignments)
        .on("mouseout", disconnect_alignments);

    assembly.svg.selectAll("rect.contig_alignment.other")
        .data(contig_alignments[contig])
        .enter()
        .append("rect")
        .attr("class", "contig_alignment other")
        .attr("contig1", contig)
        .attr("contig2", function(d) { return contig_id_name[d[2]]; })
        .attr("x", function(d) { return dim.xscale(d[3]); })
        .attr("width", function(d) { return dim.xscale(d[4])-dim.xscale(d[3]); })
        .attr("y", function(d) { return assembly.dim.yscale(contig_id_name[d[2]]) + assembly.dim.yscale.bandwidth(); })
        .attr("height", function(d) { return assembly.dim.yscale.bandwidth(); })
        .style("fill", "blue")
        .style("fill-opacity", 0.1)
        .style("stroke-opacity", 0)
        .on("mouseover", connect_alignments)
        .on("mouseout", disconnect_alignments);
}

function build_contig(contig, dim, assembly, windowsize) {

    contig.selectAll("rect.ploidy")
        .data(ploidys[contig.attr("name")])
        .enter()
        .append("rect")
        .attr("class", "ploidy")
        .attr("contig", contig.attr("name"))
        .attr("y", function(d) { return assembly.dim.yscale(contig.attr("name")); })
        .attr("height", assembly.dim.yscale.bandwidth())
        .attr("x", function(d, i) { return dim.xscale(i*windowsize); })
        .attr("width", function(d, i) {
            // either window size or the remainder at the end of the contig; subtract the left contig overhang up to position 0
            return dim.xscale(d3.min([windowsize, contig.attr("length")-i*windowsize])) - dim.xscale(0);
        })
        .attr("fill", function(d) { return ploidy_colour(assembly, d); })
        .attr("stroke-opacity", 0);

    contig.append("line")
        .attr("class", "telomere start")
        .attr("x1", dim.xscale(0))
        .attr("x2", dim.xscale(0))
        .attr("opacity", function(d) {return d.tel_start / 20; });
    
    contig.append("line")
        .attr("class", "telomere end")
        .attr("x1", function(d) { return dim.xscale(d.length) })
        .attr("x2", function(d) { return dim.xscale(d.length) })
        .attr("opacity", function(d) {return d.tel_end / 20; });
 
    contig.selectAll("line.telomere")
        .attr("contig", contig.attr("name"))
        .attr("y1", function(d) { return assembly.dim.yscale(d.name) - assembly.dim.yscale.bandwidth()/2; })
        .attr("y2", function(d) { return assembly.dim.yscale(d.name) + assembly.dim.yscale.bandwidth()*1.5; })
        .attr("style", "stroke:crimson;stroke-width:4");



}

function reorder_assembly(contigs, assembly, dim) {
    assembly.dim.yscale.domain(contigs.map(d => d.name));

    t = d3.transition().duration(1000).ease(d3.easeCubic);

    assembly.svg
        .transition(t)
        .call(d3.axisLeft(assembly.dim.yscale));

    assembly.svg.selectAll("rect.ploidy")
        .transition(t)
        .attr("y", function(d) { return assembly.dim.yscale(d3.select(this).attr("contig")); });

    assembly.svg.selectAll("line.telomere")
        .transition(t)
        .attr("y1", function(d) { return assembly.dim.yscale(d.name) - assembly.dim.yscale.bandwidth()/2; })
        .attr("y2", function(d) { return assembly.dim.yscale(d.name) + assembly.dim.yscale.bandwidth()*1.5; })

    assembly.svg.selectAll("path.contig_link")
        .transition(t)
        .attr("d", d3.linkHorizontal()
                     .x(function(d) { return d.x == -1 ? 0 : dim.xscale(d.x); })
                     .y(function(d) { return assembly.dim.yscale(d.y) + assembly.dim.yscale.bandwidth()/2; })
        );

    assembly.svg.selectAll("rect.contig_alignment.self")
        .transition(t)
        .attr("y", function(d) { return assembly.dim.yscale(d3.select(this).attr("contig1")) + assembly.dim.yscale.bandwidth(); });

    assembly.svg.selectAll("rect.contig_alignment.other")
        .transition(t)
        .attr("y", function(d) { return assembly.dim.yscale(d3.select(this).attr("contig2")) + assembly.dim.yscale.bandwidth(); });

}

function update_assembly(sort=false) {
    if (sort) {
        $('#contig_table').bootstrapTable('multipleSort');
    }
    reorder_assembly(contigs, assembly, dim);
    $('#assembly_table').bootstrapTable("load", calculate_stats(contigs));
}

function include_checkbox(value) {
    return '<input type="checkbox" ' + (value==true ? ' checked ' : '' ) + '</input>';
}

function DoCellData(cell, rowIndex, colIndex, htmlData) {
    var retVal = htmlData;
    if (htmlData.includes('checkbox')) {
        if (htmlData.includes('checked')) {
            retVal = "yes";
        } else {
            retVal = "no";
        }
    }
    return retVal;
}

var max_read_depth = 0;
var min_x_pos = -2000;
var max_x_pos = max_contig_length + 2000;

for (var key in read_alignments) { // If read_alignments is empty, this will be skipped
    max_read_depth = check_max(max_read_depth, read_alignments[key], 5);
    min_x_pos   = check_min(min_x_pos, read_alignments[key], 0);
    max_x_pos   = check_max(max_x_pos, read_alignments[key], 3);
}

if (max_read_depth == 0) { // No read alignments, so no max; but still want empty read alignment plot with ploidy levels
    max_read_depth = median_depth * 3;
}


var max_contig_name_length = d3.max(contigs.map(c => c.name.length))
var dim = { margin: {top: 100, right: 80, bottom: 70, left: d3.max([max_contig_name_length * 10, 60]) } };

dim.width = 1000 - dim.margin.left - dim.margin.right;

// Pad X axis with 2% of the maximum length, for contig alignment curves
var leftpad = (max_x_pos-min_x_pos) * 0.02
dim.xscale = d3.scaleLinear()
               .domain([min_x_pos-leftpad, max_x_pos])
               .range([0, dim.width]);


var xaxis = d3.axisBottom(dim.xscale)
              .tickFormat(function (d) {
    if (max_x_pos > 2000000) {
        return d3.format(",.1f")(d/1000000)
    }
    return d3.format(",.0f")(d)
});

assembly = build_assembly_plot(contigs, dim, windowsize, xaxis, unit);

reads = build_reads_plot(read_alignments, dim, max_read_depth, assembly, ploidy_depths, xaxis);

$.fn.editable.defaults.mode = 'inline';
$('#options_table').bootstrapTable({data: options});
$('#assembly_table').bootstrapTable({data: calculate_stats(contigs)});
$('#contig_table').bootstrapTable({
    editable: true,
    showColumns: true,
    showMultiSort: true,
    sortPriority: [{"sortName":"included","sortOrder":"desc"}, {"sortName":"length","sortOrder":"desc"}],
    showExport: true,
    exportTypes: ['csv'],
    exportOptions: { onCellHtmlData: DoCellData, fileName: outdir + ".filtered" },
    data: contigs,
    onPostBody: function() { update_assembly() },
    onCheck: function(row) {
        plot_read_alignments(read_alignments[row.name], dim, reads, contig_names[row.name], unit);
    },
    onClickCell: function(field, value, row, $element) {
        if (field === 'included') {
            row.included = value == 1 ? 0 : 1;
            update_assembly(sort=true);
        }
    }
});

update_assembly(sort=true);
$('#contig_table').bootstrapTable('check', 0);

$(".nav-item").on("click", function(){
   $(".navbar-nav").find(".active").removeClass("active");
   $(this).addClass("active");
});

$('.close').click(function() {
    $('.alert').hide();
})

$('#assembly_csv').change(function() {
    $('.alert').hide();
    csv_file = this.files[0];
    Papa.parse(csv_file, {
        header: true,
        dynamicTyping: true,
        skipEmptyLines: true,
        transformHeader: function(hf) {
            if (hf === 'Contig') return 'name';
            if (hf === 'Length') return 'length';
            if (hf === 'GC %') return 'gc';
            if (hf === 'Read Depth') return 'median_read_depth';
            if (hf === 'Group') return 'group';
            if (hf === 'Keep') return 'included';
            if (hf === 'Notes') return 'notes';
            return hf;
        },
        transform: function(value, column) {
            if (column === 'included') {
                if (value === 'yes') return 1;
                if (value === 'no') return 0;
                return true;
            };
            if (column === 'notes') {
                if (value === 'Empty') return '';
            }
            return value;
        },
        complete: function(results) {
            csv_names = results.data.map(d => d.name)
            csv_names_sorted = csv_names.slice(0).sort();
            check_contig_names = contigs.map(d => d.name).sort();
            
            // Check to make sure contig names match
            for (var i=0; i < check_contig_names.length; i++) {
                if (check_contig_names[i] !== csv_names_sorted[i]) {
                    $('.badcsv').show();
                    return;
                }
            }
            
            // Names match, so sort contigs by csv_names
            contigs = contigs.sort(function(a,b) {
                return csv_names.indexOf(a.name) - csv_names.indexOf(b.name);
            })
            
            // contigs and results now in same order, so update group, included and notes
            for (var i=0; i < results.data.length; i++) {
                contigs[i].group = results.data[i].group;
                contigs[i].included = results.data[i].included;
                contigs[i].notes = results.data[i].notes;
            }

            // Update table and assembly and show alert
            $('#contig_table').bootstrapTable("load", contigs);
            update_assembly();
            $('.goodcsv').show();
        }
    })
});
</script>
</body>
</html>
