<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="description" content="Tapestry: quality assessment for small genome assemblies">
  <meta name="author" content="Tapestry">
  
  <title>Tapestry Report</title>
  
  <style type="text/css">
      {{ include_file('static/bootstrap.min.css') }}
      {{ include_file('static/bootstrap-table.min.css') }}
  </style>
  <style>
      .xaxislabel { font-size: 12pt; font-weight: bold; }
      .yaxislabel { font-size: 12pt; font-weight: bold; }
  </style>
</head>

<body>
  <div class="mainpage">
    <h1 id="page_title">Tapestry Report</h1>

    <noscript>
      <div class="alert alert-danger">
        <h4>JavaScript Disabled</h4>
        <p>
          Tapestry uses JavaScript, but it looks like
          you have JavaScript disabled in your web browser.
          Please enable JavaScript to view the full report.
        </p>
      </div>
    </noscript>


    <div id="read_alignment_plot"></div>

    <div id="contig_alignment_plot"></div>

    <div id="assembly">
        <h2 id="assembly">Assembly</h2>
        <form id="sortvar">
            <label class="radio-inline"><input type="radio" name="sort" value="length" checked>Length</label>
            <label class="radio-inline"><input type="radio" name="sort" value="gc" >GC</label>
            <label class="radio-inline"><input type="radio" name="sort" value="cluster">Cluster</label>
        </form>
        <form id="sortdir">
            <label class="radio-inline"><input type="radio" name="sortdir" value="1">Ascending</label>
            <label class="radio-inline"><input type="radio" name="sortdir" value="-1" checked>Descending</label>
        </form>
        <div id="assembly_plot"></div>
    </div>

    <button type="button" id="export" class="btn btn-primary">Export contigs</button>

    <div id="contig_report">
      <h2>Contig Report</h2>

      <table id="table" data-sort-name="cluster" data-sort-order="asc">
        <thead>
          <tr>
            <th data-field="cluster" data-sortable="true">Cluster</th>
            <th data-field="name" data-sortable="true">Contig</th>
            <th data-field="length" data-sortable="true">Length (bp)</th>
            <th data-field="gc" data-sortable="true">GC Content (%)</th>
            <th data-field="tel_start" data-sortable="true">Start Telomeres</th>
            <th data-field="tel_end" data-sortable="true">End Telomeres</th>
          </tr>
        </thead>
      </table>
    </div>
  </div>
<script type="text/javascript">
{{ include_file('static/jquery-3.3.1.min.js') }}
{{ include_file('static/popper.min.js') }}
{{ include_file('static/bootstrap.min.js') }}
{{ include_file('static/bootstrap-table.min.js') }}
{{ include_file('static/d3.min.js') }}

var windowsize = {{ windowsize }};
var median_depth = {{ median_depth }};
var contigs = {{ contigs }};
var read_alignments = {{ read_alignments }};
var contig_alignments = {{ contig_alignments }};
var ploidys = {{ ploidys }};

sortfuncs = {
    'length'  : function(a,b) { return a.length - b.length; },
    'gc'      : function(a,b) { return a.gc - b.gc; },
    'cluster' : function(a,b) { return a.cluster - b.cluster; }
}

sortstub = function(a, b, sortvar, sortdir=1) {
    if (a.included && !b.included) return -1;
    if (b.included && !a.included) return 1;
    return sortfuncs[sortvar](a,b) * sortdir
}



var contig_names = {};
var contig_short_long = {};
contigs.forEach(function(contig) {
    contig.included = true
    contig_names[contig.longname] = contig.name;
    contig_short_long[contig.name] = contig.longname;
})

function sort_contigs (contigs, contig_positions) {
    sortvar = d3.selectAll("input[name='sort']:checked").node().value
    sortdir = d3.selectAll("input[name='sortdir']:checked").node().value
    contigs.sort(function(a,b) { return sortstub(a,b,sortvar,sortdir) });
    for (var i = 0; i < contigs.length; i++) {
        contig_positions[contigs[i].name] = i;
    }
}

var contig_positions = {};
sort_contigs(contigs, contig_positions)

ploidy_depths = [ {name:'haploid',    depth: median_depth * 0.5}, 
                  {name:'diploid',    depth: median_depth      }, 
                  {name:'triploid',   depth: median_depth * 1.5}, 
                  {name:'tetraploid', depth: median_depth * 2  },
                  {name:'repeat',     depth: median_depth * 2.5} ]

function check_max(max_val, list, i) {
    var list_max = d3.max(list, function(d) {return d[i]});
    if (max_val < list_max) {
        max_val = list_max;
    }
    return max_val;
}

function check_min(min_val, list, i) {
    var list_min = d3.min(list, function(d) {return d[i]});
    if (min_val > list_min) {
        min_val = list_min;
    }
    return min_val;
}


function mq_opacity(mq) { return mq/60 }

function plot_read_alignments(alignments, dim, assembly_dim, reads_dim, svg, contig) {

    // contig_read_alignments:
    //  0 read start including left clip
    //  1 contig alignment start
    //  2 contig alignment end
    //  3 read end including right clip
    //  4 mapping quality
    //  5 y position on plot

    svg.selectAll("text.xaxislabel").remove()
    
    svg.append("text")
       .attr("class", "xaxislabel")
       .attr("transform", "translate(" + (dim.width/2) + " ," + (reads_dim.height + dim.margin.top + 20) + ")")
       .style("text-anchor", "middle")
       .attr("fill", "black")
       .text(contig + " Position (bp)");

    svg.selectAll("line.read_alignment").remove()

    var ra = svg.selectAll("line.read_alignment")
        .data(alignments)
        .enter();
    
    ra.append("line")
        .attr("class", "read_alignment")
        .attr("x1", function(d) { return dim.xscale(d[1]); })
        .attr("x2", function(d) { return dim.xscale(d[2]); })
        .attr("y1", function(d) { return reads_dim.yscale(d[5]); })
        .attr("y2", function(d) { return reads_dim.yscale(d[5]); })
        .style("stroke-opacity", function(d) { return mq_opacity(d[4]) })
        .style("stroke", "blue");

    ra.append("line")
        .attr("class", "read_alignment")
        .attr("x1", function(d) { return dim.xscale(d[0]); })
        .attr("x2", function(d) { return dim.xscale(d[1]); })
        .attr("y1", function(d) { return reads_dim.yscale(d[5]); })
        .attr("y2", function(d) { return reads_dim.yscale(d[5]); })
        .style("stroke-opacity", function(d) { return mq_opacity(d[4]) })
        .style("stroke", "orange");

    ra.append("line")
          .attr("class", "read_alignment")
          .attr("x1", function(d) { return dim.xscale(d[2]); })
          .attr("x2", function(d) { return dim.xscale(d[3]); })
          .attr("y1", function(d) { return reads_dim.yscale(d[5]); })
          .attr("y2", function(d) { return reads_dim.yscale(d[5]); })
          .style("stroke-opacity", function(d) { return mq_opacity(d[4]) })
          .style("stroke", "pink");
}

function plot_contig_alignments(alignments, dim) {

    // contig_alignments:
    // 0 begin
    // 1 end
    // 2 y position
    // 3 contig name

    var plot_height = d3.max(alignments, function(d) { return d[2]; });
    var contigs_dim = { height: plot_height * 10 };
    
    contigs_dim.yscale = d3.scaleLinear().domain([0, plot_height])
                                 .range([0, contigs_dim.height]);
    
    var plot = d3.select("#contig_alignment_plot");

    plot.selectAll("svg").remove();

    var svg = plot.append("svg")
                  .attr("width", dim.width + dim.margin.left + dim.margin.right)
                  .attr("height", contigs_dim.height + dim.margin.top + dim.margin.bottom)
                  .append("g")
                  .attr("transform", "translate(" + dim.margin.left + ",0)")
    
    var ca = svg.selectAll("line.contig_alignment")
        .data(alignments)
        .enter();
    
    ca.append("line")
        .attr("class", "contig_alignment")
        .attr("x1", function(d) { return dim.xscale(d[0]); })
        .attr("x2", function(d) { return dim.xscale(d[1]); })
        .attr("y1", function(d) { return contigs_dim.yscale(d[2]); })
        .attr("y2", function(d) { return contigs_dim.yscale(d[2]); })
        .style("stroke", "black");
}

function plot_alignments(contig, dim, assembly_dim, reads_dim, reads_svg) {
    plot_read_alignments(read_alignments[contig], dim, assembly_dim, reads_dim, reads_svg, contig_names[contig]);
    plot_contig_alignments(contig_alignments[contig], dim);
}

function build_contig(contig, dim, assembly_dim, windowsize) {

    contig.selectAll("rect.ploidy")
        .data(ploidys[contig.attr("longname")])
        .enter()
        .append("rect")
        .attr("class", "ploidy")
        .attr("contig", contig.attr("name"))
        .attr("y", function(d) { return assembly_dim.yscale(contig.attr("name")); })
        .attr("height", assembly_dim.yscale.bandwidth())
        .attr("x", function(d, i) { return dim.xscale(i*(windowsize/2)); }) // windowsize/2 because they are sliding
        .attr("width", function(d, i) {
            // either window size or the remainder at the end of the contig; subtract the left contig overhang up to position 0
            return dim.xscale(d3.min([windowsize, contig.attr("length") - i*(windowsize/2)])) - dim.xscale(0);
        })
        .attr("fill", function(d) { return assembly_dim.ploidyscale(d/5); })
        .attr("stroke-opacity", 0);

    contig.append("circle")
        .attr("class", "telomere start")
        .attr("cx", dim.xscale(0))
        .attr("opacity", function(d) {return (d.tel_start > 0) ? 1 : 0});
    
    contig.append("circle")
        .attr("class", "telomere end")
        .attr("cx", function(d) { return dim.xscale(d.length) })
        .attr("opacity", function(d) {return (d.tel_end > 0) ? 1 : 0});

    contig.selectAll(".telomere")
        .attr("contig", contig.attr("name"))
        .attr("cy", function(d) { return assembly_dim.yscale(d.name) + assembly_dim.yscale.bandwidth()/2; })
        .attr("r", assembly_dim.yscale.bandwidth())
        .attr("fill", "red")
}

function build_stats(contig, dim, assembly_dim) {

    function ypos(d) { return assembly_dim.yscale(d.name) + assembly_dim.yscale.bandwidth()/2; }

    // Selected
    contig.append("circle")
    .attr("contig", contig.attr("name"))
    .attr("cx", 20)
    .attr("cy", function(d) { return ypos(d); })
    .attr("r", assembly_dim.yscale.bandwidth())
    .attr("fill", "none")
    .attr("stroke", "green")
    .on("click", function(d) { contig_click(d.longname); });

    // Included
    contig.append("circle")
    .attr("class", "include")
    .attr("contig", contig.attr("name"))
    .attr("cx", 40)
    .attr("cy", function(d) { return ypos(d); })
    .attr("r", assembly_dim.yscale.bandwidth())
    .attr("fill", function(d) { return contigs[contig_positions[d.name]].included ? "black" : "grey"})
    .on("click", function(d) {
        include = d3.select(this)
        contig_id = contig_positions[include.attr("contig")]
        contigs[contig_id].included = !contigs[contig_id].included
        include.attr("fill", contigs[contig_id].included ?  "black" : "grey")
        reorder_assembly()
    })

    // Cluster
    contig.append("text")
    .attr("contig", contig.attr("name"))
    .attr("x", 60 )
    .attr("y", function(d) { return ypos(d); })
    .style("fill", function(d) { return assembly_dim.clusterscale(d.cluster); })
    .style("text-anchor", "start")
    .style("alignment-baseline", "middle")
    .style("font-size", "10pt")
    .style("font-family", "monospace")
    .text(function(d) { return d.cluster });

    // GC content
    contig.append("text")
    .attr("contig", contig.attr("name"))
    .attr("x", 80 )
    .attr("y", function(d) { return ypos(d); })
    .style("fill", function(d) { return assembly_dim.gcscale(d.gc/100); })
    .style("text-anchor", "start")
    .style("alignment-baseline", "middle")
    .style("font-size", "10pt")
    .style("font-family", "monospace")
    .text(function(d) { return d.gc });
}

function reorder_assembly () {
    sort_contigs(contigs, contig_positions)
    assembly_dim.yscale.domain(contigs.map(d => d.name))

    t = d3.transition().duration(1000).ease(d3.easeCubic);

    assembly_svg
        .transition(t)
        .call(d3.axisLeft(assembly_dim.yscale))

    assembly_svg.selectAll("rect.ploidy")
        .transition(t)
        .attr("y", function(d) { return assembly_dim.yscale(d3.select(this).attr("contig")); })

    assembly_svg.selectAll("circle")
        .transition(t)
        .attr("cy", function(d) { return assembly_dim.yscale(d3.select(this).attr("contig")) + assembly_dim.yscale.bandwidth()/2; });

    stats_svg.selectAll("circle")
        .transition(t)
        .attr("cy", function(d) { return assembly_dim.yscale(d3.select(this).attr("contig")) + assembly_dim.yscale.bandwidth()/2; });

    stats_svg.selectAll("text")
        .transition(t)
        .attr("y", function(d) { return assembly_dim.yscale(d3.select(this).attr("contig")) + assembly_dim.yscale.bandwidth()/2; });
}

function select_contig(selected_contig, stats_svg, dim, assembly_dim) {
    var selected = stats_svg.selectAll("circle.selected")
                    .data(contigs.filter(c => c.longname === selected_contig), function(d) { return d.longname })
    selected.enter()
        .append("circle")
        .attr("class", "selected")
        .attr("contig", contig_names[selected_contig])
        .attr("cx", 20)
        .attr("cy", function(d) { return assembly_dim.yscale(d.name) + assembly_dim.yscale.bandwidth()/2; })
        .attr("r", assembly_dim.yscale.bandwidth())
        .attr("fill", "green");

    selected.exit().remove();
}

function contig_click(contig) {
    select_contig(contig, stats_svg, dim, assembly_dim);
    plot_alignments(contig, dim, assembly_dim, reads_dim, reads_svg);
}

var max_read_depth = 0;
var min_read_pos = 0;
var max_read_pos = 0;

for (var key in read_alignments) {
    max_read_depth = check_max(max_read_depth, read_alignments[key], 5);
    min_read_pos   = check_min(min_read_pos, read_alignments[key], 0);
    max_read_pos   = check_max(max_read_pos, read_alignments[key], 3);
}

var max_cluster = d3.max(contigs.map(c => c.cluster))

var dim = { margin: {top: 100, right: 80, bottom: 50, left: 80} };

dim.width = 1000 - dim.margin.left - dim.margin.right;

dim.xscale = d3.scaleLinear()
               .domain([min_read_pos-1000, max_read_pos]) // -1000 to give axis some space
               .range([0, dim.width]);

xaxis = d3.axisBottom(dim.xscale);


var assembly_dim = {height: contigs.length * 20};

assembly_dim.yscale = d3.scaleBand()
                      .domain(contigs.map(d => d.name))
                      .rangeRound([0, assembly_dim.height])
                      .padding(0.8);

assembly_dim.ploidyscale = d3.scaleSequential(d3.interpolateGreens);
assembly_dim.gcscale = d3.scaleSequential(d3.interpolateRainbow);
assembly_dim.clusterscale = d3.scaleOrdinal(d3.schemeDark2);

var assembly_svg = d3.select("#assembly_plot")
            .append("svg")
            .attr("width", dim.width + dim.margin.left + dim.margin.right/2)
            .attr("height", assembly_dim.height + dim.margin.top + dim.margin.bottom)
            .append("g")
            .attr("transform", "translate(" + dim.margin.left + "," + (dim.margin.top + assembly_dim.height) + ")")
            .call(xaxis)
            .attr("class", "xaxis")
            .append("g")
            .attr("transform", "translate(0,-" + assembly_dim.height + ")")
            .call(d3.axisLeft(assembly_dim.yscale))
            .attr("class", "yaxis");

assembly_svg
    .selectAll(".tick")
    .on("click", function(d) { contig_click(contig_short_long[d]); })

var stats_svg = d3.select("#assembly_plot")
                .append("svg")
                .attr("width", 100 + dim.margin.right)
                .attr("height", assembly_dim.height + dim.margin.top + dim.margin.bottom)
                .append("g")
                .attr("transform", "translate(0," + dim.margin.top + ")")

stats_svg.selectAll("text")
    .data(["Selected", "Included", "Cluster", "GC"])
    .enter()
    .append("text")
    .attr("transform", function(d, i) { return "translate(" + (((i+1) * 20) + 5) + ",0) rotate(-55)"; })
    .attr("fill", "black")
    .style("text-anchor", "start")
    .style("font-size", "10pt")
    .text(function(d) { return d; });


var reads_dim = { height: 400 };

reads_dim.yscale = d3.scaleLinear()
                     .domain([0, max_read_depth])
                     .range([reads_dim.height, 0]);

yaxis = d3.axisLeft(reads_dim.yscale)
                     
reads_svg = d3.select("#read_alignment_plot")
              .append("svg")
              .attr("width", dim.width + dim.margin.left + dim.margin.right)
              .attr("height", reads_dim.height + dim.margin.top + dim.margin.bottom)
              .append("g")
              .attr("transform", "translate(" + dim.margin.left + "," + (dim.margin.top + reads_dim.height) + ")")
              .call(xaxis)
              .attr("class", "xaxis")
              .append("g")
              .attr("transform", "translate(0,-" + reads_dim.height + ")")
              .call(yaxis)
              .attr("class", "yaxis")

reads_svg.append("text")
         .attr("class", "yaxislabel")
         .attr("transform", "rotate(-90)")
         .attr("y", 0 - (dim.margin.left * 0.6))
         .attr("x",0 - (reads_dim.height / 2))
         .attr("fill", "black")
         .style("text-anchor", "middle")
         .text("Reads"); 

reads_svg.selectAll("line.ploidy")
   .data(ploidy_depths.filter(d => d.name != "repeat"))
   .enter()
   .append("line")
   .attr("class", "ploidy")
   .attr("x1", dim.xscale.range()[0]+1) // +1 to avoid axis
   .attr("x2", dim.xscale.range()[1])
   .attr("y1", function(d) { return reads_dim.yscale(d.depth) })
   .attr("y2", function(d) { return reads_dim.yscale(d.depth) })
   .style("stroke", function(d, i) { return assembly_dim.ploidyscale(i/5) })
   .style("stroke-width", 2)

reads_svg.selectAll("text.ploidy")
   .data(ploidy_depths)
   .enter()
   .append("text")
   .attr("x", dim.xscale.range()[1] + 10 )
   .attr("y", function(d) { return reads_dim.yscale(d.depth)+2 })
   .style("fill", function(d, i) { return assembly_dim.ploidyscale(i/5) })
   .style("text-anchor", "start")
   .text(function(d) { return d.name })


var selected_contig = contigs.map(d => d.longname)[0];
plot_alignments(selected_contig, dim, assembly_dim, reads_dim, reads_svg);

var contig_groups = assembly_svg.selectAll("g.contig")
    .data(contigs, function(d) { return d.name })
    .enter()
    .append("g")
    .attr("class", "contig")
    .attr("name", function(d) { return d.name })
    .attr("longname", function(d) { return d.longname })
    .attr("length", function(d) { return d.length })
    .on("click", function(d) { contig_click(d.longname); })
    .each(function(d) { build_contig(d3.select(this), dim, assembly_dim, windowsize); });

var stats = stats_svg.selectAll("g.stat")
    .data(contigs, function(d) { return d.name })
    .enter()
    .append("g")
    .attr("class", "stat")
    .attr("name", function(d) { return d.name })
    .attr("longname", function(d) { return d.longname })
    .each(function(d) { build_stats(d3.select(this), dim, assembly_dim); });

select_contig(selected_contig, stats_svg, dim, assembly_dim);

d3.selectAll("input[name='sort'], input[name='sortdir']").on("change", function() {reorder_assembly()})

$('#export').click(function (e) {
    var text = contigs.filter(c => c.included).map(c => c.name).join('\n');
    var filename = "filtered_assembly.txt";

    var element = document.createElement('a');
    element.setAttribute('href', 'data:text/plain;charset=utf-8,' + encodeURIComponent(text));
    element.setAttribute('download', filename);
    element.style.display = 'none';
    document.body.appendChild(element);
    element.click();
    document.body.removeChild(element);
})

$(function() {
  $('#table').bootstrapTable({data: contigs})
});

</script>
</body>
</html>